### PR TITLE
Drive modes, persistent params, new robot model support.

### PR DESCRIPTION
**Pull request by jbrover**

### CODE REVIEW

File: include/control.hpp [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+  void setOutputLimits(pid_output_limits pid_output_limits);
+  pid_output_limits getOutputLimits();
+
+  void setIntegralErrorLimit(float error_limit);
+  float getIntegralErrorLimit();
+
+  pid_outputs runControl(float target, float measured);
+
+  void writePidDataToCsv(std::ofstream &log_file, pid_outputs data);
```
Comment (drhieu) [inline include/control.hpp]: /* brief
*
*/ are missing for these function. Intellisense would not be able to disaply any useful description for these

File: include/control.hpp [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+  float getOutputDecay();
+
+  void setOpenLoopMaxRpm(float open_loop_max_motor_rpm);
+  float getOpenLoopMaxRpm();
+
+  void setTrim(float left_trim, float right_trim);
+  float getLeftTrim();
+  float getRightTrim();
+
+  void setAngularScaling(angular_scaling_params angular_scaling_params);
```
Comment (drhieu) [inline include/control.hpp:407]: missing /*brief*/ for these as well

File: src/control.cpp [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+#ifdef DEBUG
+  /*open a log file to store control data*/
+  auto t = std::time(nullptr);
+  auto tm = *std::localtime(&t);
+
+  std::ostringstream oss;
+  oss << std::put_time(&tm, "%d-%m-%Y-%H-%M-%S");
+  auto filename = oss.str();
+  std::cerr << "log file name " << filename + ".csv" << std::endl;
+  log_file_.open("/home/rover/Documents/" + filename + ".csv");
```
Comment (drhieu) [inline src/control.cpp:335]: *magic* file path

File: tools/pid_visualizer.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+data = pd.read_csv(args.filename)
+
+# left pid
+left_pid_data = data[data["name"] == "pid_left"]
+
+# right pid
+right_pid_data = data[data["name"] == "pid_right"]
+
+# motion data
+motion_data = data[data["type"] == "motion"]
```
Comment (drhieu) [inline tools/pid_visualizer.py:73]: will break on Independent wheel mode

**Review by drhieu — COMMENTED** [review]
missing brief statements for some .hpp

**Review by drhieu — COMMENTED** [review]
PID_visualizer will not work on independent wheel; but this is for internal use so its ok ?

File: include/control.hpp (hunk 1 above)
Comment (jbrover) [reply include/control.hpp]: everything should be filled in now.

File: tools/pid_visualizer.py (hunk 4 above)
Comment (jbrover) [reply tools/pid_visualizer.py:73]: for internal use, OK for now.

**Review by drhieu — APPROVED** [review]

**Pull request merged**